pkg/hubble/relay/pool/manager.go

```diff
@@ -0,0 +1,361 @@
+			}).Warning("Failed to create peer notify client for peers change notification; will try again after the timeout has expired")
+			select {
+			case <-m.stop:
+				cancel()
+				return
+			case <-time.After(m.opts.RetryTimeout):
+				continue
+			}
+		}
+		for {
+			select {
+			case <-m.stop:
+				cl.Close()
+				cancel()
+				return
+			default:
+			}
+			cn, err := client.Recv()
+			if err != nil {
+				cl.Close()
+				m.opts.Log.WithFields(logrus.Fields{
+					"error":              err,
+					"connection timeout": m.opts.RetryTimeout,
+				}).Warning("Error while receiving peer change notification; will try again after the timeout has expired")
+				select {
+				case <-m.stop:
+					cancel()
+					return
+				case <-time.After(m.opts.RetryTimeout):
+					continue connect
+				}
+			}
+			m.opts.Log.WithField("change notification", cn).Debug("Received peer change notification")
+			p := hubblePeer.FromChangeNotification(cn)
+			switch cn.GetType() {
+			case peerpb.ChangeNotificationType_PEER_ADDED:
+				m.add(p)
+			case peerpb.ChangeNotificationType_PEER_DELETED:
+				m.remove(p)
+			case peerpb.ChangeNotificationType_PEER_UPDATED:
+				m.update(p)
+			}
+		}
+	}
+}
+
+func (m *Manager) manageConnections() {
+	for {
+		select {
+		case <-m.stop:
+			return
+		case name := <-m.offline:
+			m.mu.Lock()
+			p := m.peers[name]
+			m.mu.Unlock()
+			// a connection request has been made, make sure to attempt a connection
+			m.connect(p, true)
+		case <-time.After(m.opts.ConnCheckInterval):
+			var retry []*peer
+			m.mu.Lock()
+			now := time.Now()
+			for _, p := range m.peers {
+				p.mu.Lock()
+				if p.conn != nil {
+					switch p.conn.GetState() {
+					case connectivity.Connecting, connectivity.Idle, connectivity.Ready, connectivity.Shutdown:
+						p.mu.Unlock()
+						continue
+					}
+				}
+				if p.nextConnAttempt.IsZero() || p.nextConnAttempt.Before(now) {
+					retry = append(retry, p)
+				}
+				p.mu.Unlock()
+			}
+			m.mu.Unlock()
+			for _, p := range retry {
+				m.connect(p, false)
+			}
+		}
+	}
+}
+
+// Stop implements PeerManager.Stop.
+func (m *Manager) Stop() {
+	close(m.stop)
+}
+
+// List implements PeerManager.List.
+func (m *Manager) List() []Peer {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if len(m.peers) == 0 {
+		return nil
+	}
+	peers := make([]Peer, 0, len(m.peers))
+	for _, v := range m.peers {
+		// note: there shouldn't be null entries in the map
+		peers = append(peers, Peer{
+			Peer: hubblePeer.Peer{
+				Name:    v.Name,
+				Address: v.Address,
+			},
+			Conn: v.conn,
+		})
+	}
+	return peers
+}
+
+// ReportOffline implements PeerManager.ReportOffline.
+func (m *Manager) ReportOffline(name string) {
+	go func() {
+		m.mu.Lock()
+		p, ok := m.peers[name]
+		m.mu.Unlock()
+		if !ok {
+			return
+		}
+		p.mu.Lock()
+		if p.conn != nil {
+			switch p.conn.GetState() {
+			case connectivity.Connecting, connectivity.Idle, connectivity.Ready:
+				// it looks like it's actually online or being brought online
```

Comment:
How do we know that? What guarantees do we have that this go routine was not concurrently scheduled with a 
 `m.connect(p, ...)`?

Reply:
It may have been but it doesn't really matter; this is more a sanity check to ensure the caller does not report offline a peer that is indeed online (or about to become) to avoid unnecessary work. Also, there's a lock on `p`.